<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
	
	<script src="libs/fineuploader-3.8.2/jquery-2.0.3.js"></script>
	<script type="text/javascript" src="libs/bootstrap/js/bootstrap.js"></script>
	<script src="libs/fineuploader-3.8.2/jquery.fineuploader.js"></script>
	<!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
      <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
	<script src="http://cdnjs.cloudflare.com/ajax/libs/snap.svg/0.3.0/snap.svg-min.js"></script>
    <script src="//cdnjs.cloudflare.com/ajax/libs/noUiSlider/6.2.0/jquery.nouislider.min.js"></script>
	
	
	<link rel="stylesheet" type="text/css" href="libs/fineuploader-3.8.2/fineuploader.css" >
	<link rel="stylesheet" type="text/css" href="libs/fineuploader-3.8.2/fineuploader-custom.css" >
	<link rel="stylesheet" type="text/css" href="libs/flatui/bootstrap/css/bootstrap.css">
	<link rel="stylesheet" type="text/css" href="libs/flatui/bootstrap/css/bootstrap-custom.css">
	<link rel="stylesheet" type="text/css" href="libs/flatui/css/flat-ui.css">
	<link rel="stylesheet" type="text/css" href="libs/flatui/css/flatui-custom.css">
	<link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/noUiSlider/6.2.0/jquery.nouislider.min.css">

	
<title>CandiSNP</title>
	
</head>
<body>


	<div style="max-width: 960px">
	<header class="header-custom">
		 <h1 class="tab" ><span style="color:#34BEDA">Candi</span><span style="color:#FFB416">SNP:</span><span style="color:#34BEDA"> identifying candidate SNPs in genomes</span></h1>
	</header>
	</div>
	
	<hr width=1000 align=left>
	<div style="max-width: 960px">
	<h4 align="right" style="color:#8E8F98">version 0.2.0 - <i>"beau jour blaireau"</i></h4>
	</div>
	<div class="div-text">
	<h4><span style="color:#34BEDA">Candi</span><span style="color:#FFB416">SNP</span> analyses and provides per-chromosome plots of SNPs across genomes. It provides the location of the SNPs and uses colour-coded palettes to show if the SNPs are in coding or non-coding regions and if they are synonymous or non-synonymous.</h4>
	</div>
	<hr width=1000 align=left>
	
	<a class="btn btn-info margin-left=10" id="PageRefresh">
      Reset data fields
    </a>
	
	<script type="text/javascript">
 
	$('#PageRefresh').click(function() {
    	      location.reload();
	});
 
</script>
	<div class="div-custom">
	
		<h3>1. Select a genome <img  style="width:20px;" class="pop" src="img/info.png" alt="Info" rel="popover" data-content="<h4>Select the genome to be analysed.</h4>" data-original-title="Genome selection"/></h3>
		
		<select  id="genome_selection" class="tip" title="Select your genome from which your SNP data was called">
			<option value="athalianaTair10">Arabidopsis TAIR10</option>
			<option value="athalianaTair9">Arabidopsis TAIR9</option>
			<option value="rice7">Rice genome v7</option>
			<option value="SL2.40">Tomoto genome v2.40</option>
			<option value="gmax1.09v8">Glycine max genome 1.09v8</option>
			<option value="grape">Grape genome V1</option>
			<option value="maizeZmB73">Maize B73</option>
		</select>
		
		<br>
		<h4 id="filter_cen_txt">Filter out centromere snps?</h4>
		<select  id="filter_cen" class="tip" title="Filter out snps 1Mb either side of centromeres">
			<option value="yes">Yes</option>
			<option value="no">No</option>
		</select>
		</div>
		
		<script>
		$('#genome_selection').change(function(){    
			if($(this).val() == "athalianaTair9" || $(this).val() == "athalianaTair10"){
				$('#filter_cen_txt').fadeIn();
				$('#filter_cen').fadeIn();
			}
			else{
				$('#filter_cen_txt').hide();
				$('#filter_cen').hide();
			}
		});
		</script>
		
		
		<div class="div-custom">
			<h3>2. Select SNP data <img style="width:20px;" class="pop" src="img/info.png" alt="Info" rel="popover"
			data-content="<h4>Your data file must contain a header line with columns for 'Chr', 'Pos', 'Ref', 'Alt' and 'Allele_Freq'. They can be in any order. 
			Chromosome names are important. Read the 'Help' section at the bottom of the page for exact details of the format you must use.</h4>" data-original-title="CandiSNP SNP Data"/></h3>
			
			<div id="manual-fine-uploader">
			
		</div>
		<div id="sliderHelperTxt"></div>
		<div id="slider"></div>
			
	</div>

	
	

	<div class="div-custom">
	<h3>3. Select palette to view SNPs <img  style="width:20px;" class="pop" src="img/info.png" alt="Info" rel="popover" data-content="<h4>Choose the palette to pick the colour of the dots on the plots</h4>" data-original-title="Palette selection"/></h3>
	<select  id="palette_selection" class="tip" title="The colour scheme for representing your data">
		<option>CandiSNP</option>
		<option>contrast</option>
		<option>diverse</option>
		<option>gradient</option>
		<option>diverging</option>
		<option>sequential</option>
	</select>
	
	</div>
	<div class="div-custom">
	<h3>4. Upload and analyse </h3>
	<div id="triggerUpload" class="btn btn-primary" style="margin-top: 10px;">
		<a class="tip" style="color: #ffffff;" title="Your files will be uploaded to the server and analysed">
		<i class="icon-upload icon-white" ></i>Analyse</a>
	</div>
	</div>
	<hr>
	
	
	<div class="div-help">
	<h3>Help </h3>
	<div>
	<h4>1. Selecting a genome</h4>
	There are a number of built-in genomes included in <span style="color:#34BEDA">Candi</span><span style="color:#FFB416">SNP</span>, which are necessary for analysing SNPs.
	The genome databases for TAIR10, TAIR9, Rice genome v7, Tomoto genome v2.40, Glycine max genome 1.09v8, Grape genome V1 and Maize B73 are currently available.
	New genomes can be added as required. <p>
	It is often helpful to filter out SNPs around the centromere. If you select either of the Arabidopsis genomes, you will have the option to filter out SNPs 1Mb either side of the centromere
	</p>
	</div>
	
	<hr>
	
	<div>
	<h4>2. Selecting SNP data</h4>
	
	<span style="color:#34BEDA">Candi</span><span style="color:#FFB416">SNP</span> is designed to be as flexible as possible
	with input data. It requires data in a file saved in the comma-delimited (or '.csv') format, which can be done in Excel. The data must have the following column headers:<p>
	'Chr', 'Pos', 'Ref', 'Alt' and 'Allele_Freq'. Other column names may be present in the data - they will just be ignored.<p>
	It's essential that the chromosome names in column 'Chr' match that in the database. Here are examples of chromosome names for each build:<p>
	
	<table border="1">
	<tr><th>Genome</th><th>Chromosome names</th></tr>
	<tr><td>TAIR10</td><td>1, 2, 3, 4, 5, M, C</td></tr>
	<tr><td>TAIR9</td><td>1, 2, 3, 4, 5, M, C</td></tr>
	<tr><td>Rice genome v7</td><td>1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, Un, Sy</td></tr>
	<tr><td>Tomato genome v2.40</td><td>SL2.40ch00, SL2.40ch01, SL2.40ch02, SL2.40ch03, SL2.40ch04, SL2.40ch05, SL2.40ch06, SL2.40ch07, SL2.40ch08, SL2.40ch09, SL2.40ch10, SL2.40ch11, SL2.40ch12</td></tr>
	<tr><td>Glycine max genome 1.09v8</td><td>Gm01, Gm02, Gm03, Gm04, Gm05, Gm06, Gm07, Gm08, Gm09, Gm10, Gm11, Gm12, Gm13, Gm14, Gm15, Gm16, Gm17, Gm18, Gm19, Gm20 and 21 to 2287</td></tr>
	<tr><td>Grape genome</td><td>1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 1_random, 10_random, 11_random, 12_random, 13_random, 16_random, 17_random, 18_random, 3_random, 4_random, 5_random, 7_random, 9_random</td></tr>
	<tr><td>Maize B73</td><td>1, 4, 2, 3, 5, 7, 8, 6, 9, 10, UNKNOWN, Mt, Pt</td></tr>
	</table>
	<br>Once you have selected the data to upload, an Allele Frequency box will appear next to each file (with a default value of 0.7). This value can be changed before you submit your data for analysis. It can also be changed later on in the analysis too.
	</div>
	<hr>
	<div>
	<h4>3. The Palette</h4>
	<span style="color:#34BEDA">Candi</span><span style="color:#FFB416">SNP</span> offers the user a choice of four different palletes.
	The colours represent SNPs on the <span style="color:#34BEDA">Candi</span><span style="color:#FFB416">SNP</span> plots.
	The palettes you can choose are<p>
	
	<table border="1" cellpadding="5">
	<tr ><th>Palette name</th><th>Non-Synonymous<br>in Coding Region C-T or G-A</th><th>Non-Synonymous<br>in Coding Region</th><th>Synonymous<br>in Coding Region</th><th>Non Coding Region</th></tr>
	<tr><td height="15">CandiSNP</td><td bgcolor="#808080"></td><td bgcolor="#808080"></td><td bgcolor="#FF0000"></td><td bgcolor="#808080"></td></tr>
	<tr><td height="15">gradient</td><td bgcolor="#225EA8"></td><td bgcolor="#41B6C4"></td><td bgcolor="#A1DAB4"></td><td bgcolor="#FFFFCC"></td></tr>
	<tr><td height="15">contrast</td><td bgcolor="#D7191C"></td><td bgcolor="#FDAE61"></td><td bgcolor="#2C7BB6"></td><td bgcolor="#ABD9E9"></td></tr>
	<tr><td height="15">diverging</td><td bgcolor="#D01C8B"></td><td bgcolor="#F1B6DA"></td><td bgcolor="#B8E186"></td><td bgcolor="#4DAC26"></td></tr>
	<tr><td height="15">sequential</td><td bgcolor="#E31A1C"></td><td bgcolor="#FD8D3C"></td><td bgcolor="#FECC5C"></td><td bgcolor="#FFFFB2"></td></tr>
	<tr><td height="15">diverse</td><td bgcolor="#CB181D"></td><td bgcolor="#FB6A4A"></td><td bgcolor="#FCAE91"></td><td bgcolor="#FEE5D9"></td></tr>
	</table>
	</div>
	
	<hr>
	<div>
	<h4>4. Analysis</h4>
	Once you click the 'Analyse' button, your data is uploaded to the <span style="color:#34BEDA">Candi</span><span style="color:#FFB416">SNP</span> server
	where it is analysed. The result of the analysis will be a plot which will be displayed on the screen and a data file providing annotations of the SNPs.
	The image and the data file can be saved by clicking on the 'Save image' or 'Save data' buttons.<p>
	Re-analysis using a different allele frequency can be carried out by changing the 'Allele frequency' value and clicking 'Resubmit'. This can be done any number of times for a range of any different allele frequencies.
	After each resubmission the data and image can be saved.	
	</div>
	</div>
	
	<script>
	$(document).ready(function() {
		
		//when the upload button is clicked, upload the files
		$('#triggerUpload').click(function() {
			//get the number of stored files waiting to be uploaded
			var storedFiles = $('#manual-fine-uploader').fineUploader('getUploads', { status: qq.status.SUBMITTED }).length
			//we'll presume for now they're all good
			var allGood = true;
			//loop through each file
			for ( var i = 0; i < storedFiles; i = i + 1 ){
				//clear any previous warnings
				$('#processor'+i).empty();
				//get each allele frequency in turn
				var current_allele_freq = $('#manual-fine-uploader').find('input[name="allele_freq'+i+'"]').val();
				//if it's not a valid allele frequency, set allGood to false and show a warning
				if (!$.isNumeric(current_allele_freq) || current_allele_freq > 1)
				{
					allGood = false;
					$('#processor'+i).append('<p style="color:white;background-color:red;font-size:20px;"> Insert a number between 0 and 1</p>');
				}
			}
			
			//we've finished checking the allele frequencies
			//if they're all good upload all files
			if (allGood == true)
			{
				$('#manual-fine-uploader').fineUploader('uploadStoredFiles');
			}

		});
		
		var manualuploader = $('#manual-fine-uploader').fineUploader({
			request: {
				endpoint: 'cgi/upload.cgi',
				timeout: 600000
			},
			autoUpload: false,
			
			text: {
			uploadButton: '<i class="icon-plus icon-white"></i> Select or drop Files'
		  }
			
		})
		
		//called when the file or Blob has been successfully submitted to the uploader. 
		//each time a file is opened insert a text box next to it for the allele frequency
		//and give it a default value of 0.9
		.on('submitted', function(event, id, name) {
			//console.log('Event: ' + event + ' id: ' + id +' name: ' + name);
			var fileItemContainer = $(this).fineUploader('getItemByFileId', id);
			$(fileItemContainer).append('Allele frequency: <input class="span1" title="Enter an allele frequency between 0-1" type="text" name="allele_freq'+id+'" value="0.7" >');
			$(fileItemContainer).append('<div id="resubmit'+id+'"></div>')
			$(fileItemContainer).append('<div id="processor'+id+'"></div>')
		})
		
		//called just before a file or Blob upload begins.
		.on('upload', function(event, id, name) {
			//get the identify the current file container
			var fileItemContainer = $(this).fineUploader('getItemByFileId', id),
			//get the entered allele frequency
			enteredAlleleFreq = $(fileItemContainer).find('input[name="allele_freq'+id+'"]').val();
			var palette_selected = $("#palette_selection option:selected").val();
			var genome_selected = $("#genome_selection option:selected").val();
			var filter_cent_snps = "no";
			if (genome_selected == "athalianaTair9" || genome_selected == "athalianaTair10") {
				filter_cent_snps = $("#filter_cen option:selected").val();

			}
			console.log("filter - "+filter_cent_snps);
			$(this).fineUploader('setParams', {allele_freq: enteredAlleleFreq, genome: genome_selected, palette: palette_selected, filter_snps: filter_cent_snps}, id);
			//console.log('enteredAlleleFreq:' + enteredAlleleFreq + " genome: " + genome_selected + " palette: " + palette_selected + " id: " + id);
		})
		
		//called when the file or Blob upload has finished
		//when the initial upload has finished create and show a resubmit and 'save data' button and load the plot
		.on('complete', function(event, id, fileName, responseJSON) {
			$('#resubmit'+id).append('<input type="submit" id="resubmitButton'+id+'"value="Resubmit" class="btn btn-primary" title="Re-analyse data with new allele frequency"/>');
			
			//when the resubmit button is pressed, get the new allele frequency
			$('#resubmitButton'+id).click(function() {
				var current_allele_freq = $(fileItemContainer).find('input[name="allele_freq'+id+'"]').val();
				var palette_selected = $("#palette_selection option:selected").val();
				var genome_selected = $("#genome_selection option:selected").val();
				var filter_cent_snps = "no";
				if (genome_selected == "athalianaTair9" || genome_selected == "athalianaTair10") {
					filter_cent_snps = $("#filter_cen option:selected").val();
				}
				//check that it's numeric and less of equal to one
				//if it's not, alert the user
				if (!$.isNumeric(current_allele_freq) || current_allele_freq >1)
				{
					//console.log('Allele freq:' + current_allele_freq );
					alert('Use a number between 0 and 1 for file '+fileName);
				}
				//otherwise, process the file
				else
				{
					//show the 'processing' icon whilst it's working
					$('#processor'+id).append("<img src='libs/fineuploader-3.8.2/processing.gif' >");
					
					//and post the new allele freq, along with the filename (which will already be on the server, so
					//no need to re-upload the file) and run the resubmit.cgi script
					$.ajax({
						type: 'POST',
						url: 'cgi/resubmit.cgi',
						async: true,
						data: {allele_freq: current_allele_freq, file_name : fileName, genome: genome_selected, palette: palette_selected, filter_snps: filter_cent_snps}, 
						dataType: 'json',
						//if everything runs OK and 'success' is returned... 
						success: function(data) {
							//display the new plot
							$('#imageLoc'+id).empty().append('<img class="slide-svg" src="public/'+data.file+'.svg"></img>');
							//remove the 'processing' icon
							$('#processor'+id).empty();
							var baseFileName = fileName.substring(0, fileName.lastIndexOf('.') );
							console.log("base file name: "+baseFileName);
							$('#saveData'+id).empty().append('<a href="public/'+data.file+'.csv" download="'+baseFileName+'.csv" class="btn btn-small btn-info" title="Save your data to file">Save data</a>');
							$('#saveData'+id).append('<a href="public/'+data.file+'.svg" download="'+baseFileName+'.svg" margin="10px" class="btn btn-small btn-info" title="Save the plot to file">Save image</a>');
						}
					});
				}
			})
			
			console.log ("current file name "+responseJSON.file);
			
			var fileItemContainer = $(this).fineUploader('getItemByFileId', id);
			//add the Resubmit and save data buttons and div for the image 
			$(fileItemContainer).append('<div id="saveData'+id+'"></div>');
			$(fileItemContainer).append('<div id="imageLoc'+id+'"></div>');
			$(fileItemContainer).append('<div id="imageLoc'+id+'"></div>');
			//if everything runs OK
			if (responseJSON.success) {
				var baseFileName = fileName.substring(0, fileName.lastIndexOf('.') );
				console.log ("base file name "+baseFileName);
				//use the html5 'download' attribute to download the svg
				//(which would just be rendered to screen by browser using the filedowload method above)
				$('#saveData'+id).append('<a href="public/'+responseJSON.file+'.csv" download="'+baseFileName+'.csv" class="btn btn-small btn-info" title="Save your data to file">Save data</a>');
				$('#saveData'+id).append('<a href="public/'+responseJSON.file+'.svg" download="'+baseFileName+'.svg" class="btn btn-small btn-info" margin="10px" title="Save the plot to file">Save image</a>');
				//append the image to the div
				$('#imageLoc'+id).append('<img class="slide-svg" src="public/'+responseJSON.file+'.svg"></img>');
				refreshSlider();
			}
			$(fileItemContainer).append('<hr>');
		});
		
	});
	</script>

<script>
$(function () {
	$(".pop").popover(
	{
		trigger: 'hover',
		html: 'true',
		delay: 500,
		content: '<div style="font-size:14px;">'
	});
});
</script>
<script>
$(function () {
    $('.tip').tooltip()
});
</script>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-43984900-1', 'tsl.ac.uk');
  ga('send', 'pageview');

</script>
<!--<script>-->
<!--      refreshSlider=function() {-->
<!--	  setTimeout(function() {-->
<!--	  console.log('refreshed');-->
<!--	  -->
<!--        var a = document.getElementsByClassName("slide-svg");-->
<!--        // Get the SVG document inside the Object tag-->
<!--        $("#slider").noUiSlider({-->
<!--          start: 1,-->
<!--          range: {-->
<!--            'min': 0.4,-->
<!--            'max': 5-->
<!--          }-->
<!--        });-->
<!--		-->
<!--		var helperText = '<h4>Move slider to change scale of dots on plot</h4>';-->
<!--		-->
<!--		$('#sliderHelperTxt').empty().append(helperText);-->
<!--		-->
<!---->
<!--        $('#slider').change(function(){-->
<!--			console.log('Change');-->
<!--			-->
<!--			-->
<!--			$.each(a, function(itt,aa){-->
<!--					var doc = aa.contentDocument;-->
<!--					var surface = doc.getElementById("surface1");  -->
<!--					var points = surface.getElementsByTagName("path");-->
<!--					var scale = $("#slider").val();-->
<!--		-->
<!--					-->
<!--				  $.each(points,function(i, point){-->
<!--					if(!point.style.strokeMiterlimit && point.style.fill != "rgb(255, 255, 255)"){-->
<!--					  var s = Snap(point);-->
<!--					  var t = new Snap.Matrix();-->
<!--					  point.getBBox();-->
<!--					  var x = point.getBBox().x + (point.getBBox().width/2);-->
<!--					  var y = point.getBBox().y + (point.getBBox().height/2);-->
<!--					  t.scale(scale, scale, x, y); -->
<!--					  s.transform(t);-->
<!--					}-->
<!--				  });-->
<!--				});-->
<!--        });-->
<!--		}, 1000);-->
<!--      }-->
<!--	  -->
<!--      </script>-->
<br>
<div class="div-contact">
	Questions or comments? Contact: dan.maclean _at_ tsl.ac.uk</div>
</body>


</html>


